### PR TITLE
gh-125159: Speed up `posixpath.realpath()` in the common case

### PR DESCRIPTION
**Pull request by nineteendo**
Originally proposed in #120146.

### Benchmark

<details><summary>Show script</summary>

```sh
# patch-1.sh
mkdir -p "tmp/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a"

echo 200 .
main/python.exe    -m timeit -s "import os; path = './' * 200"   "os.path.realpath(path)"
patch-1/python.exe -m timeit -s "import os; path = './' * 200"   "os.path.realpath(path)"

echo 200 ..
main/python.exe    -m timeit -s "import os; path = '../' * 200"   "os.path.realpath(path)"
patch-1/python.exe -m timeit -s "import os; path = '../' * 200"   "os.path.realpath(path)"

echo 200 dirs
main/python.exe    -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/a' * 200" "os.path.realpath(path)"
patch-1/python.exe -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/a' * 200" "os.path.realpath(path)"
```

</details>

<details><summary>200 . - no difference</summary>

```none
10000 loops, best of 5: 26.9 usec per loop
10000 loops, best of 5: 26.7 usec per loop
```

</details>

<details><summary>200 .. - 1.02x faster </summary>

```none
5000 loops, best of 5: 50.3 usec per loop
5000 loops, best of 5: 49.2 usec per loop
```

</details>

<details><summary>200 dirs - no difference</summary>

```none
100 loops, best of 5: 2.19 msec per loop
100 loops, best of 5: 2.16 msec per loop
```

</details>

```shell
$ python -m timeit -s "name = '.'" "pass" "if name is None: pass"
50000000 loops, best of 5: 9.03 nsec per loop
$ python -m timeit -s "name = '.'" "pass"
50000000 loops, best of 5: 6.08 nsec per loop
```

* Issue: gh-125159

### CODE REVIEW

**Comment by nineteendo** [comment]
cc @barneygale

**Comment by ericvsmith** [comment]
Does this make a measurable difference?

File: Lib/posixpath.py [hunk 1]
```diff
@@ -428,13 +428,14 @@ def _realpath(filename, strict=False, sep=sep, curdir=curdir, pardir=pardir,
...
         if not name or name == curdir:
-            # current dir
-            continue
+            if name is None:
+                # resolved symlink target
+                seen[rest.pop()] = path
+                continue
+            else:
+                # current dir
+                continue
```
Comment (barneygale) [inline Lib/posixpath.py]:
```suggestion
            # current dir
            if name is None:
                # resolved symlink target
                seen[rest.pop()] = path
            continue
```

**Comment by Wulian233** [comment]
Could you provide the results of the speed test?

**Comment by nineteendo** [comment]
It's not a lot faster, 2.95 nsec/iteration.

**Comment by ericvsmith** [comment]
I think this change isn't worth the churn.

**Comment by nineteendo** [comment]
Just as expected.

**Comment by erlend-aasland** [comment]
> Just as expected.

If I understand you correctly, you expected this PR to be rejected.

The next time you consider opening a PR that you "expect to be rejected", please consider again the following:

- is it worth the reviewer and CI churn to post a "ready-to-be-rejected" PR?
- is there anything you can improve _post_ proposing the PR that can increase the chances of it being accepted? (benchmarks, seeking consensus about the change, backwards compatibility, etc.) See also the devguide: [Making good PRs](https://devguide.python.org/getting-started/pull-request-lifecycle/#making-good-prs)

**Comment by barneygale** [comment]
@erlend-aasland it's my fault - see https://github.com/python/cpython/pull/120146#discussion_r1792511985

@nineteendo really sorry for giving you bad advice and wasting your time

**Comment by nineteendo** [comment]
Thanks for being honest, Barney, but it's also my fault. I knew it wasn't a lot faster (much less than the 5% Alex mentioned in the past) and therefore probably wasn't worth a PR to change just that. That doesn't mean that making the code 3 nanoseconds slower for speeding up a rare case is fine though.

**Comment by vstinner** [comment]
In general, an optimization is worth it if it's at least 10% faster on a microbenchmark. It doesn't seem to be the case here.

**Pull request closed without merging**